### PR TITLE
More stages for more plugins

### PR DESCRIPTION
**Pull request by yuval-k**
suggestion to allow more plugins stages in gloo

### CODE REVIEW

File: projects/gloo/pkg/plugins/plugin_interface.go [hunk 1]
```diff
@@ -108,13 +108,31 @@ type StagedHttpFilter struct {
 
 type FilterStage int
 
+const Space = 100
+
+const (
+	Fault FilterStage = iota*Space*3 + Space + Space/2
```
Comment (yuval-k) [inline projects/gloo/pkg/plugins/plugin_interface.go]: should we add CorsStage here?

Comment (mitchdraft) [reply projects/gloo/pkg/plugins/plugin_interface.go]: what about calling it "pre auth" - as in, not just for cors

Comment (yuval-k) [reply projects/gloo/pkg/plugins/plugin_interface.go]: i think we can have a stage to each of our filters. so we'll have more flexibility

**Comment by mitchdraft** [comment]
instead of stretching iotas, how about a mapping function?

```diff
git diff projects/gloo/pkg/translator/
diff --git a/projects/gloo/pkg/translator/http_filters.go b/projects/gloo/pkg/translator/http_filters.go
index b6498cd8..489859e8 100644
--- a/projects/gloo/pkg/translator/http_filters.go
+++ b/projects/gloo/pkg/translator/http_filters.go
@@ -76,18 +76,18 @@ func (t *translator) computeHttpFilters(params plugins.Params, listener *v1.Http
        }

        // sort filters by stage
-       envoyHttpFilters := sortFilters(httpFilters)
+       envoyHttpFilters := t.sortFilters(httpFilters)
        envoyHttpFilters = append(envoyHttpFilters, &envoyhttp.HttpFilter{Name: envoyutil.Router})
        return envoyHttpFilters
 }

-func sortFilters(filters []plugins.StagedHttpFilter) []*envoyhttp.HttpFilter {
+func (t *translator) sortFilters(filters []plugins.StagedHttpFilter) []*envoyhttp.HttpFilter {
        // sort them first by stage, then by name.
        less := func(i, j int) bool {
                filteri := filters[i]
                filterj := filters[j]
-               if filteri.Stage != filterj.Stage {
-                       return filteri.Stage < filterj.Stage
+               if t.stageMap(filteri.Stage) != t.stageMap(filterj.Stage) {
+                       return t.stageMap(filteri.Stage) < t.stageMap(filterj.Stage)
                }
                return filteri.HttpFilter.Name < filterj.HttpFilter.Name
        }
diff --git a/projects/gloo/pkg/translator/translator.go b/projects/gloo/pkg/translator/translator.go
index 86ec76df..a0a2681c 100644
--- a/projects/gloo/pkg/translator/translator.go
+++ b/projects/gloo/pkg/translator/translator.go
@@ -20,19 +20,30 @@ type Translator interface {
        Translate(params plugins.Params, proxy *v1.Proxy) (envoycache.Snapshot, reporter.ResourceErrors, error)
 }

+type StageMapFunc func(stage plugins.FilterStage) int
+
 type translator struct {
        plugins             []plugins.Plugin
        extensionsSettings  *v1.Extensions
        settings            *v1.Settings
        sslConfigTranslator utils.SslConfigTranslator
+       stageMap            StageMapFunc
+}
+
+func DefaultStageMap(stage plugins.FilterStage) int {
+       return int(stage)
 }

-func NewTranslator(sslConfigTranslator utils.SslConfigTranslator, settings *v1.Settings, plugins ...plugins.Plugin) Translator {
+func NewTranslator(sslConfigTranslator utils.SslConfigTranslator, settings *v1.Settings, stageMap StageMapFunc, plugins ...plugins.Plugin) Translator {
+       if stageMap == nil {
+               stageMap = DefaultStageMap
+       }
        return &translator{
                plugins:             plugins,
                extensionsSettings:  settings.Extensions,
                settings:            settings,
                sslConfigTranslator: sslConfigTranslator,
+               stageMap:            stageMap,
        }
 }
```

**Pull request merged**